### ISSUE TITLE
reinterpret_cast and undefined behaviour

### ISSUE THREAD

**Issue by lcarlier** [lefticus/cpp_weekly#129]
**Channel**

C++Weekly

**Topics**

reinterpret_cast and undefined behaviour.
What is allowed and not allowed.
i.e. can we cast integer to pointer (e.g. for register access)? any risk doing so?
type aliasing.
Some example demonstrating what can go wrong when wrong type aliasing is done.

**Length**

5-10 minutes

**Comment by lefticus** [comment]
Reference to existing episode on this topic: https://youtu.be/L06nbZXD2D0

**Comment by lcarlier** [comment]
Hi Jason,
Thanks! That explains already a lot!

I was wondering whether there is also undefined behaviour when doing volatile reinterpret cast when accessing HW registers.

i.e.

```c++
uintptr_t CONFIG_REG_ADDR = 0x10000000;
uint32_t config_reg_val = *reinterpret_cast<volatile uint32_t*>(CONFIG_REG_ADDR);
```

Can we conclude that whenever data alignment is guaranteed, then `reinterpret_cast` won't produce UB?

Note that I've seen that using memcpy in this [case](https://godbolt.org/z/9xjr1E1a4) gives the same ASM with optimized build but I'm still wondering if all the code I've seen so far is just full of UB or if we were just lucky.

**Comment by carlosgalvezp** [comment]
@lcarlier In this case you have a integer-to-pointer conversion, which as per the [Standard](https://eel.is/c++draft/expr.reinterpret.cast#5) is _implementation defined_. Most likely all microcontroller toolchains implement this correctly as it's the only way to access memory-mapped devices.

**Comment by Dharmesh946** [comment]
see also https://github.com/lefticus/cpp_weekly/issues/245#issuecomment-2220224938
links with std::bit_cast, std::start_lifetime_as,...

**Comment by Dharmesh946** [comment]
> Reference to existing episode on this topic: https://youtu.be/L06nbZXD2D0

1- Hi, how is it possible to get guarantees that memcpy is optimized away?
2- After the second memcpy (back to the storage), now S object exists in the storage and subsequent reinterpret_cast would be valid?
3- how to use that for non default constructible objects? (I'm thinking about allocating the temporary with std::allocator::allocate and destroying it at the end of the function, but it won't be optimized away: https://godbolt.org/z/6zbTMEPnj)

Regards

**Comment by VincentZalzal** [comment]
With the [latest episode (512)](https://youtu.be/JtFVyXQ00PQ?si=5dBA3PWDEG1-W6cn) about `std::start_lifetime_as`, I wonder : what are the remaining legitimate uses of `reinterpret_cast`? If I parsed cppreference correctly, and ignoring some more obscure uses:
- `T*` to `std::uintptr_t` and back
- internal pointer to unrelated opaque handle pointer and back (as it is never dereferenced while it is a handle)
- `T*` to `std::byte*` (or `char`, `unsigned char`) to see its representation, and back
- accessing a contiguous sequence of signed integers as unsigned integers of the same size (or the other way around). Also works for a single integer, but I would use `static_cast` for that.
- function pointer to `void*` and back (for `dlsym`)

**Comment by Dharmesh946** [comment]
@VincentZalzal : creating objects inside suitably aligned storage and getting them back:

you have an array of bytes "storage"
you place-new some object of type T there (respecting alignment, see std::align)
you retrieve a pointer to a T through std::launder(reinterpret_cast<T*>(storage+offset)), offset being the offset in byte inside the storage to your object.

to avoid that you can save the result of the placement new to a T* variable.

**Comment by VincentZalzal** [comment]
@Dharmesh946 You're right! I see this as a special case of point 3 (`T*` to `std::byte*` and back), but it is a useful case, especially when implementing containers.

**Comment by Dharmesh946** [comment]
@VincentZalzal I would say that examining object representation is another matter. It is now bugged in the standard (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2025/p1839r7.html). The best way I found was to use std::bit_cast the object to examine to an array of bytes, hoping for the compiler to recognize the pattern and optimize.
converting pointer from and to integer has no operational value, only for debugging as this conversion is compiler dependent.
I think that the last 2 points are better achieved through static_cast.

**Comment by Dharmesh946** [comment]
another use-case for `reinterpret_cast`: with `std::construct_at`. AFAIU it requires a location as `T*` so if you want to construct a T on, say, a storage of bytes:
```cpp
alignas(T) std::byte storage[sizeof(T)];
T* ptr = std::construct_at(reinterpret_cast<T*>(storage),args);
```
this is legal as the storage address is assumed to be voidified internally (https://stackoverflow.com/a/77233807/21691539)

I hope it will be addressed also there: #509 

side note, I'm wondering why the signature of  `std::construct_at`would not have directly taken a `void*`

**Comment by carlosgalvezp** [comment]
> would not have directly taken a void*

Exactly. It feels like placement new is a much cleaner solution already, no casts required.

**Comment by lcarlier** [comment]
https://en.cppreference.com/w/cpp/memory/construct_at.html

Construct_at is constexpr and consteval in C++26

**Comment by LB--** [comment]
> Construct_at is constexpr and consteval in C++26

The problem is that `std::construct_at` has to forward its arguments, so it does not work for in-place construction of aggregates, especially with types that are not movable nor copyable.

> It feels like placement new is a much cleaner solution already, no casts required.

You always have to explicitly cast to `void*` to make sure you actually select placement new and not some other operator new overload.

**Comment by Dharmesh946** [comment]
Regarding `std::construct_at` signature, an uneducated afterthought: the `T*` is maybe to have template argument deduction. But if it"s that, IMHO, it's a poor design choice. `std::construct_at<T>(void_ptr,args)` would be much cleaner and explicit than requiring an evil reinterpret_cast. By the way, the same could be said of `std::launder`. If both function add to be called with an explicit template parameter but a `void *` argument I have no more use case for `reinterpret_cast`

**Comment by Dharmesh946** [comment]
another use case that I missed is function pointer cast to another function pointer type. The latter can be called in a defined way if the type is call-compatible with the original function definition (https://eel.is/c++draft/expr.call#5).
Basically, it is a way to allow some erasure type for function: you reinterpret_cast it's address to some common type (void(*)() for instance) then back to it's original type as needed.

**Comment by LB--** [comment]
If you're going to use `reinterpret_cast` anyway, I find it preferable to use `std::uintptr_t` for stuff like that, rather than worrying about the specifics of call compatibility. Just be sure to always cast back to exactly the same original type and you're fine. Also, `reinterpret_cast<std::uintptr_t>` is required for properly checking if pointers are within buffers in a defined way, since it's otherwise UB to compare pointers that aren't from the same range. The standard tiptoes around this with `std::less` sometimes though, mainly for sorting.

**Comment by Dharmesh946** [comment]
@LB-- I actually don't find any usage of `reinterpret_cast<std::uintptr_t>` other than debugging purpose. Do you think of one?
By limiting the number of legit usages (I referenced 3 so far). It would be easy to wrap them in some utility function, so that user never write it anymore. For instance:
- retrieve_address_as<T*> == std::launder(reinterpret_cast<T*>)
- my_construct_at<T>(void*,...) == construct_at(T*,...)
- retrieve_function_as<ret_t,args_t...>(void(*)()) == reinterpret_cast<ret_t(*)(args_t...)>
(sorry for the pseudo-code)

**Comment by LB--** [comment]
I'm not sure what you mean, I literally just gave an example in my prior comment where  `reinterpret_cast<std::uintptr_t>` is required to avoid UB. You can certainly wrap it in a helper function, but that doesn't make it not required.

**Comment by Dharmesh946** [comment]
@LB-- Sorry I don't get your example. Might you share a compiler explorer snippet?

**Comment by LB--** [comment]
I can do you one better: here is a detailed write-up on exactly why `reinterpret_cast<std::uintptr_t>` is required in some cases: https://devblogs.microsoft.com/oldnewthing/20170927-00/?p=97095

**Comment by Dharmesh946** [comment]
Thanks @LB--. I see now what you thought about.
Yet, "The pointer-to-integer conversion is implementation-defined, which means that your implementation must document how it works. If your implementation defines the pointer-to-integer conversion as producing the numeric value of the linear address of the object referenced by the pointer, and you know that you are on a flat architecture," => it's completely implementation dependent. Therefore I don't add it to my list of use cases. If legitimate on a given architecture, I would encapsulate it inside a function dully identified, usable only on this architecture and hiding the reinterpret_cast.
